### PR TITLE
ENH: add the File Manager and Memeory buffers manager classes

### PR DESCRIPTION
**Pull request by awalter-bnl**
This PR adds the File Manager and Memory buffers manager classes to suitcase.utils. It is expected that these will be used in the suitcase.* export and serializer functions.

### CODE REVIEW

**Comment by awalter-bnl** [comment]
Added this PR at the request of @danielballan

File: suitcase/utils/__init__.py [hunk 1]
```diff
@@ -1,25 +1,249 @@
...
+        errors : string or None
+            Passed through to open. See Python open documentation for allowed
+            values.
+        Returns
+        -------
+        file : handle
+        """
+        filepath = self.reserve_name(label, postfix)
+        # create the directories if they don't yet exist
+        if not os.path.exists(os.path.dirname(filepath)):
```
[the remark this answers is not in the document]
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: fixed

File: suitcase/utils/__init__.py [hunk 2]
```diff
@@ -1,25 +1,249 @@
...
+
+    def close(self):
+        '''close all files open by the manager
+        '''
+        for f in self._files:
+            f.close()
+
+
+class PersistentStringIO(io.StringIO):
+    ''' A StringIO that does not clear the buffer when closed or excited from
```
[the remark this answers is not in the document]
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: fixed

File: suitcase/utils/__init__.py [hunk 3]
```diff
@@ -1,25 +1,249 @@
...
+        elif mode == 'xb':
+            buffer = PersistentBytesIO()
+        else:
+            raise ModeError(
+                f'the mode passed to MemoryBuffersWrapper.open is {mode} but '
+                'needs to be one of "x", "xt" or "xb"')
+        self.buffers[postfix] = buffer
+        return buffer
+
+    def close(self):
```
[the remark this answers is not in the document]
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: This was added at the express wish of @tacaswell (he asked for a close on both the manager and the serializer). I do not understand your argument above, as long as `manager.close()` is not called it should not do anything I think?

[the remark this answers is not in the document]
Comment (danielballan) [reply suitcase/utils/__init__.py]: Could there be a misunderstanding? I think that `export` will call `serializer.close()` ([example](https://github.com/NSLS-II/suitcase-tiff/pull/1/files#diff-8ccbf61f6f17a225be10116a327c4595R114)) and then, as written here, that will erase the data. I think @tacaswell, you, and I all agreed that both `manager` and `Serializer` should implement a `close()` method. But this particular `close()` method should do nothing, just as `PersistentStringIO.close()` does nothing.

File: suitcase/utils/__init__.py [hunk 4]
```diff
@@ -241,6 +234,7 @@ def open(self, label, postfix, mode, encoding=None, errors=None):
         self.buffers[postfix] = buffer
         return buffer
 
+
```
Comment (danielballan) [inline suitcase/utils/__init__.py]: Also -- a small thing -- flake8 will complain about this line. :-D

File: suitcase/utils/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tacaswell) [reply suitcase/utils/__init__.py]: I think @danielballan has it right, both levels need close, but this should call `.close` on it's buffers (or just be a no-op).

[the remark this answers is not in the document]
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: yeah @danielballan and I spoke I just haven't had time to update it yet

[the remark this answers is not in the document]
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: resolved

File: suitcase/utils/__init__.py (hunk 4 above)
Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: sorry, forgot to pass it through flake8 before commiting, I have resolved all the flake8 issues in the latest commit

File: suitcase/utils/__init__.py [hunk 5]
```diff
@@ -1,25 +1,243 @@
...
+        return buffer
+
+    def close(self):
+        '''close all buffers open by the manager
+
+        Note: this is a no-op method added for API consistency only, for the
+        details on why it is a no-op see Persistent*IO definitions in
+        `suitcase.utils.__init__`
+        '''
+        ...
```
Comment (danielballan) [inline suitcase/utils/__init__.py]: If you take a second look at @tacaswell's follow-up comment, you'll see he is suggesting that this should call `close()` on all its buffers, even though that will have no effect. I hadn't thought of this in my original suggestion, but I agree; if later `Persistent*IO` is changed or more likely swapped out for something else by another develop with a use case that has not occurred to us, _two_ layers of no-op may confuse them.

Comment (awalter-bnl) [reply suitcase/utils/__init__.py]: fix made

**Comment by danielballan** [comment]
Activated Travis-CI for this repo and power-cycled to trigger it.

**Comment by awalter-bnl** [comment]
I don't think there are any tests in this repo? I suppose it will at least run flake8!

**Comment by danielballan** [comment]
100% test coverage!

**Pull request merged**